rfcs/0001-inline-expression-value-logging.md

```diff
@@ -0,0 +1,305 @@
+# Summary
+
+Add support to `log` for inline expression and value logging, as a
+superset of the `dbg!` macro recently added to rust `std`.
+
+# Motivation
+
+The motivation is in part the same as for the accepted `dbg!` macro of
+[RFC 2361], as implemented in rust 1.32. To summarize:
+
+* It is convenient to be able to insert logging into larger expressions
+  without needing to restructure using additional `let` bindings, or
+  by duplicating sub-expressions.
+
+* For debug/trace level logging in particular, automatic formatting of
+  the expression with its value can give sufficient context in the log
+  (particularly with _target_, _file_ and _line_ information) and avoids
+  additional boilerplate.
+
+In a project where configuring `log` and an output logger
+implementation hasn't (yet) happened, one can conveniently use
+[`std::dbg!`](https://doc.rust-lang.org/std/macro.dbg.html), with no
+initial setup cost, for `expression = value` printing to STDERR. From
+the linked rustdoc:
+
+> Note that the macro is intended as a debugging tool and therefore
+> you should avoid having uses of it in version control for longer
+> periods. Use cases involving debug output that should be added to
+> version control may be better served by macros such as `debug!` from
+> the `log` crate.
+
+Indeed, a major point of the `log` package and `Level`s, is the
+ability to keep `debug!` and `trace!` logging in place for further
+use, including by other contributors, without paying a cost for
+unlogged messages in release builds.
+
+It follows that for projects that _do already have_ `log` and an
+output logger dependency and configuration, and particularly with some
+debug/trace logging already in place, use of `std::dbg!` would be
+unwelcome in PRs, and likely less productive than a `log`-based
+equivalent.
+
+## `std::dbg!` doesn't work well in projects with `log` configured
+
+In the context of unit tests, when using `cargo test`, _stderr_ (or
+_stdout_) configured log output isn't captured, while `dbg!`, at least
+when used within the tests, is captured.  For passing tests, only log
```

Comment:
I would like to understand this better before using it as motivation. How is it that a logger configured to stderr or stdout circumvents the test harness's capture of stderr and stdout? Is that a bug that needs to be fixed, eliminating the motivation?

There is value in successful tests remaining quiet.

Reply:
This might be specifically directed at `env_logger`, which uses `termcolor` to write to the terminal by default. `termcolor` in turn doesn't use the `print!` or `eprint!` macros so it doesn't hook into some special unstable APIs those macros and `libtest` use to figure out if output should be captured. We're working around it in `env_logger` at the moment, but I'd call this a point-in-time issue.

Reply:
My observations are with the `fern` logger implementation, which I chose for its features with a CLI binary, and then figured would be best to stick with for test logging.  It ultimately uses `write!(std::io::Stderr,…)` I believe.   So it seems that the test harness isn't actually capturing stderr/out at the process level. Perhaps it can't without also abandoning test threads in a single process.  It seems pretty fragile.  Will it even work when `dbg!` or `eprinln!` are found in lib code, outside of test cfg?

Reply:
Digging around a bit more, I found this summary of the capture weakness of libtest pretty helpful:
 
https://github.com/rust-lang/rust/issues/50297#issuecomment-388988381

It suggests it is also vulnerable when other threads beyond the test thread log.

Reply:
Refined this motivation section in the RFC to link to the same and better describe the issue, and its awkward workarounds.